include/aws/common/math.h

```diff
@@ -91,10 +91,10 @@ AWS_STATIC_IMPL size_t aws_mul_size_saturating(size_t a, size_t b) {
 }
 
 AWS_STATIC_IMPL int aws_mul_size_checked(size_t a, size_t b, size_t *r) {
-    if ((uint64_t)SIZE_MAX == ~(uint32_t)0) {
-        return (int)aws_mul_u32_checked((uint32_t)a, (uint32_t)b, (uint32_t *)r);
+    if (sizeof(size_t) == sizeof(uint32_t)) {
```

Comment:
There should be a way to do this at compile time, right? Can you do:
```c
#if SIZE_MAX == UINT32_MAX
    return aws_mul_u32_checked(a, b, r);
#elif SIZE_MAX == UINT64_MAX
    return aws_mul_u64_checked(a, b, r);
#else
#   error "Target not supported"
#endif
```

Reply:
Yes.  That looks good.

Reply:
> (of course, the test will need to no-op out on 32-bit systems)

What's the best way to make a test that only runs on some systems? Is there a flag in the testing framework?

Reply:
Also, should I apply that format to the ```aws_mul_size_saturating()``` function, while I'm here? Or save that for a separate PR?

Reply:
1. Just `#ifdef` the body of the test, so on 32 bits the only code that is compiled is `return AWS_OP_SUCCESS`
2. Yeah I'd say go for it, if we say "let's do it later" it'll never happen 😛

Reply:
Weirdly, it seems to require the cast: if I remove it, I get
```
make[2]: *** [CMakeFiles/aws-c-common.dir/source/byte_buf.c.o] Error 1
In file included from /Users/dsn/ws/CCommon-CBMC/aws-c-common/source/date_time.c:17:
In file included from /Users/dsn/ws/CCommon-CBMC/aws-c-common/include/aws/common/array_list.h:19:
/Users/dsn/ws/CCommon-CBMC/aws-c-common/include/aws/common/math.h:97:38: error: incompatible pointer types passing 'size_t *' (aka 'unsigned long *') to parameter of type 'uint64_t *' (aka 'unsigned long long *') [-Werror,-Wincompatible-pointer-types]
    return aws_mul_u64_checked(a, b, r);
                                     ^
/Users/dsn/ws/CCommon-CBMC/aws-c-common/include/aws/common/math.gcc_overflow.inl:39:75: note: passing argument to parameter 'r' here
AWS_STATIC_IMPL int aws_mul_u64_checked(uint64_t a, uint64_t b, uint64_t *r) {
                                                                          ^
1 error generated.
make[2]: *** [CMakeFiles/aws-c-common.dir/source/date_time.c.o] Error 1
```

Reply:
If it's trying to tell you that a `size_t` is 32 bits, then something's wrong with the branching.

Reply:
I did some digging.  It seems that on MacOS, ```size_t``` is a 64 bit ```long unsigned int``` while ```uint64_t``` is a 64bit ```unsigned long long```.  The compiler complains when you try to turn a pointer to one into a pointer to the other without a cast, even tho they are the same size.